### PR TITLE
Add support for Midonet plugin

### PR DESCRIPTION
**Pull request by mmnelemane**
Add support for Midonet plugin for neutron and facilitate selection of Midonet as one of the possible networking plugins from crowbar UI for neutron. Customer already has a patch on Cloud6 without crowbar_framework integration. This is a forward-port of the patch for integration into Cloud7 product.

### CODE REVIEW

File: chef/cookbooks/neutron/templates/default/neutron_lbaas.conf.erb [hunk 1]
```diff
@@ -74,7 +74,11 @@ region_name = <%= @keystone_settings['endpoint_region'] %>
 # Combination of <service type> and <name> must be unique; <driver> must also be unique
 # This is multiline option
 # service_provider=LOADBALANCER:name:lbaas_plugin_driver_path:default
-# service_provider=LOADBALANCER:Haproxy:neutron_lbaas.services.loadbalancer.drivers.haproxy.plugin_driver.HaproxyOnHostPluginDriver:default
+<% if @networking_plugin != "midonet" -%>
+service_provider=LOADBALANCER:Haproxy:neutron_lbaas.services.loadbalancer.drivers.haproxy.plugin_driver.HaproxyOnHostPluginDriver:default
+<% else -%>
+service_provider=LOADBALANCER:Midonet:midonet.neutron.services.loadbalancer.driver.MidonetLoadbalancerDriver:default
+<% end -%>
```
Comment (vuntz) [inline chef/cookbooks/neutron/templates/default/neutron_lbaas.conf.erb]: The service_provider is set some lines later, so this is not the right location anymore.

File: chef/data_bags/crowbar/migrate/neutron/049_add_midonet_attributes.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+def upgrade(ta, td, a, d)
+  a["midonet"] = {}
+  a["midonet"]["host"] = ta["midonet"]["host"]
+  a["midonet"]["port"] = ta["midonet"]["port"]
+  a["midonet"]["username"] = ta["midonet"]["username"]
+  a["midonet"]["password"] = ta["midonet"]["password"]
+  a["midonet"]["openstack_user"] = ta["midonet"]["openstack_user"]
+  a["midonet"]["openstack_password"] = ta["midonet"]["openstack_password"]
```
Comment (vuntz) [inline chef/data_bags/crowbar/migrate/neutron/049_add_midonet_attributes.rb]: You can simply do:
```ruby
unless a.key?("midonet")
  a["midonet"] = ta["midonet"]
end
```

File: crowbar_framework/app/assets/javascripts/barclamps/neutron/application.js [hunk 3]
```diff
@@ -253,6 +254,19 @@ function networking_plugin_check() {
     $('#cisco_switches').hide();
     $('#cisco_ports').hide();
```
Comment (vuntz) [inline crowbar_framework/app/assets/javascripts/barclamps/neutron/application.js]: You should have `$('#midonet_container').hide();` here (in the second line in the `case`, not at the end)

File: crowbar_framework/app/assets/javascripts/barclamps/neutron/application.js [hunk 4]
```diff
@@ -253,6 +254,19 @@ function networking_plugin_check() {
...
     break;
+  case 'midonet':
```
Comment (vuntz) [inline crowbar_framework/app/assets/javascripts/barclamps/neutron/application.js]: Missing `$('#vmware_container').hide();`

File: crowbar_framework/config/locales/neutron/en.yml [hunk 5]
```diff
@@ -117,5 +124,6 @@ en:
         cisco_nexus_ovs: 'The "cisco_nexus" mechanism driver needs also the "openvswitch" mechanism driver'
         openvswitch_linuxbridge: 'The "openvswitch" and "linuxbridge" mechanism drivers cannot be used in parallel. Only select one of them'
         dvr_vmware: 'DVR is not compatible with VMware NSX plugin'
+        dvr_midonet: 'DVR is not compatible with Midokura Midonet plugin'
```
Comment (vuntz) [inline crowbar_framework/config/locales/neutron/en.yml]: I'm not sure we should mention Midokura, as Midonet is an open source project.

File: crowbar_framework/app/views/barclamp/neutron/_edit_attributes.html.haml [hunk 6]
```diff
@@ -40,6 +40,14 @@
         = string_field %w(vmware tz_uuid)
         = string_field %w(vmware l3_gw_uuid)
 
+      #midonet_container
+        = string_field %w(midonet host )
```
Comment (vuntz) [inline crowbar_framework/app/views/barclamp/neutron/_edit_attributes.html.haml]: Remove spurious space before `)`

File: chef/cookbooks/neutron/templates/default/neutron_lbaas.conf.erb (hunk 1 above)
Comment (vuntz) [inline chef/cookbooks/neutron/templates/default/neutron_lbaas.conf.erb]: Btw, I'm unsure if Midonet supports LBaaSv2. If not, then the code around https://github.com/crowbar/crowbar-openstack/pull/558/files#diff-dc28d262bd8f17de96133ca22b6ccc6fR104 needs some change.

**Comment by rhafer** [comment]
@mmnelemane Sorry, I don't think I'll be able to look into this anytime soon :(. So if you need a timely review, you might want to ask somebody else.

**Comment by mmnelemane** [comment]
@vuntz: Re-review ?

**Comment by matelakat** [comment]
This change has to be rebased.

File: chef/cookbooks/neutron/attributes/default.rb [hunk 7]
```diff
@@ -108,6 +108,7 @@
     cisco_opflex_pkgs: ["agent-ovs",
                         "lldpd",
                         "openstack-neutron-opflex-agent"],
+    midonet_pkgs: ["python-networking-midonet"],
```
Comment (matelakat) [inline chef/cookbooks/neutron/attributes/default.rb]: What is the source of this package? I cannot find it. Isn't it `openstack-networking-midonet` ?

File: chef/data_bags/crowbar/migrate/neutron/049_add_midonet_attributes.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+def upgrade(ta, td, a, d)
```
Comment (matelakat) [inline chef/data_bags/crowbar/migrate/neutron/049_add_midonet_attributes.rb]: If we want this patch to work for Cloud7, migration revision has to be something like 1xx

File: chef/cookbooks/neutron/attributes/default.rb (hunk 7 above)
Comment (aplanas) [inline chef/cookbooks/neutron/attributes/default.rb]: We have this:

https://build.opensuse.org/project/show/home:aplanas:midokura

Maybe is the `openstack-neutron-midonet` one.

**Comment by matelakat** [comment]
Closing down this PR, and re-opening as #585 that'll be owned by matelakat

File: chef/cookbooks/neutron/attributes/default.rb (hunk 7 above)
Comment (matelakat) [inline chef/cookbooks/neutron/attributes/default.rb]: @aplanas I'm pretty sure it's the package you pointed at. Do we have any source for documentation? How is the customer supposed to install this? There must be some sort of manual, like add repository xyz or retrieve binaries xyz. Any idea where that can be?

**Pull request closed without merging**